### PR TITLE
Add rules-objstore initial implementation

### PR DESCRIPTION
**Pull request by onprem**
This is the initial implementation of Object Storage based Rules API backend (more details [here](https://observatorium.io/docs/proposals/20201019-prometheus-rules-in-observatorium-api.md/)) for Observatorium API.

### CODE REVIEW

File: .golangci.yml [hunk 1]
```diff
@@ -0,0 +1,28 @@
+run:
+  # default concurrency is a available CPU number
+  concurrency: 4
+  # timeout for analysis, e.g. 30s, 5m, default is 1m
+  deadline: 3m
+  tests: true
+
+linters:
+  disable:
```
[the remark this answers is not in the document]
Comment (onprem) [reply .golangci.yml:9]: Nope, just picked them from opa-openshift project. I have enabled them now.

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+ARCH ?= $(shell uname -m)
+
+VERSION := $(strip $(shell [ -d .git ] && git describe --always --tags --dirty))
+BUILD_DATE := $(shell date -u +"%Y-%m-%d")
+BUILD_TIMESTAMP := $(shell date -u +"%Y-%m-%dT%H:%M:%S%Z")
+VCS_BRANCH := $(strip $(shell git rev-parse --abbrev-ref HEAD))
+VCS_REF := $(strip $(shell [ -d .git ] && git rev-parse --short HEAD))
+DOCKER_REPO ?= quay.io/observatorium/rules-objstore
+
+default: rules-objstore
```
[the remark this answers is not in the document]
Comment (onprem) [reply Makefile]: Done!

File: pkg/server/server.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
+package server
+
+import (
+	"io"
+	"net/http"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	rulesspec "github.com/observatorium/api/rulesbackend/server/v1"
```
[the remark this answers is not in the document]
Comment (onprem) [reply pkg/server/server.go]: Using the implementation merged in main now.

File: pkg/server/server.go [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+	"github.com/thanos-io/objstore"
+)
+
+const (
+	rulesPath = "/rules.yaml"
+)
+
+type Server struct {
```
[the remark this answers is not in the document]
Comment (onprem) [reply pkg/server/server.go:17]: I think it's generally a good practice to export a type used outside the package (even if it is indirectly used). We have a variable of type `Server` in our main package.

[the remark this answers is not in the document]
Comment (jessicalins) [reply pkg/server/server.go:17]: ah yes, then it makes sense! :+1:

File: go.mod [hunk 5]
```diff
@@ -1,3 +1,20 @@
 module github.com/observatorium/rules-objstore
 
-go 1.17
+go 1.16
```
[the remark this answers is not in the document]
Comment (onprem) [reply go.mod]: Reverted to `1.17`.

File: pkg/server/server.go [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+	if _, err := io.Copy(w, file); err != nil {
+		level.Warn(logger).Log("msg", "copying rules file to HTTP response", "err", err)
+	}
+}
+
+func (s *Server) SetRules(w http.ResponseWriter, r *http.Request, tenant string) {
+	logger := log.With(s.logger, "handler", "setrules", "tenant", tenant)
+
+	err := s.bucket.Upload(r.Context(), tenant+rulesPath, r.Body)
```
[the remark this answers is not in the document]
Comment (jessicalins) [reply pkg/server/server.go]: yeah I wonder if we would have the case where `tenant+rulesPath` does not exist (therefore we can't set/list rules)
so the question is if it does make sense to validate this somehow or just fallback to the error returned from `s.bucket.Upload(..)` 
actually for listing rules we appear to already do that 
https://github.com/observatorium/rules-objstore/blob/8a9d196802e7214e30d22173399328bb3a3c56e6/pkg/server/server.go#L36

[the remark this answers is not in the document]
Comment (onprem) [reply pkg/server/server.go]: Ah, got it. The way Upload works is that it overwrites an object if it already exists. This is also what we want to do in `PUT` handler so we don't need to validate if it already exists.

File: main.go [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+				http.MethodGet,
+				http.StatusNotFound,
+				time.Second,
+			),
+		)
+	}
+
+	bkt, err := client.NewBucket(logger, cfg.BucketConfig, reg, cfg.Name)
+	if err != nil {
+		stdlog.Fatalf("creating object store bucket client: %v", err)
```
Comment (bwplotka) [inline main.go:68]: Ideally you have only fatals/panics in `main` and dedicate `run() error` function, but not a biggy

**Review by bwplotka — APPROVED** [review]
Pretty good for start. Thanks!

File: go.mod [hunk 8]
```diff
@@ -1,3 +1,91 @@
 module github.com/observatorium/rules-objstore
 
 go 1.17
+
+replace (
+	// TODO: Remove this: https://github.com/thanos-io/thanos/issues/3967.
+	github.com/minio/minio-go/v7 => github.com/bwplotka/minio-go/v7 v7.0.11-0.20210324165441-f9927e5255a6
+
+	github.com/thanos-io/thanos v0.22.0 => github.com/thanos-io/thanos v0.19.1-0.20211112050938-aa7e9f33aaf0
```
[the remark this answers is not in the document]
Comment (onprem) [reply go.mod:9]: Going with thanos directly for now. I thought it'd be better to work on the objstore lib effort in parallel on not get blocked on this.

**Pull request merged**